ci/infra/testrunner/tests/utils.py

```diff
@@ -0,0 +1,34 @@
+import signal
+import time
+
+def wait(func, *args, **kwargs):
+
+    class TimeoutError(Exception):
+        pass
+
+    timeout = kwargs.pop("wait_timeout", 5)
+    delay   = kwargs.pop("wait_delay", 0)
+    backoff = kwargs.pop("wait_backoff", 1)
+    retries = kwargs.pop("wait_retries", 3)
+    allow   = kwargs.pop("wait_allow", ())
+    def _handle_timeout(signum, frame):
+        raise TimeoutError()
+
+    time.sleep(delay)
+    reason=""
+    for i in range(0, retries):
+        signal.signal(signal.SIGALRM, _handle_timeout)
+        signal.alarm(timeout)
```

Comment:
This solution won't allow for nesting wait calls (e.g. calling `wait` when the callback `func` calls `wait` again on its own). Creating a thread and using `thread.wait(timeout=<timeout>)` would be safer if we expect this scenario.

Reply:
Good point. Can we document this limitation of nested waits and implement the thread in a follow up PR?

Reply:
Absolutely, resolving.